include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+#include "revng/ADT/UpcastablePointer.h"
+#include "revng/ADT/UpcastablePointer/YAMLTraits.h"
+#include "revng/Model/TupleTree.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+
+namespace flatc {
+
+namespace TypeKind {
+
+/// \brief Enum for identifying different kind of flatc types.
+enum Values {
+  Invalid,
+  Primitive,
+  Enum,
+  Typedef,
+  Struct,
+  Union,
+  FunctionPtr,
+};
+// TODO: fromString toString
+} // namespace TypeKind
+
+struct Type;
+struct PrimitiveType;
+struct EnumType;
+struct TypedefType;
+struct StructType;
+struct UnionType;
+struct FunctionPtrType;
+
+struct TypesTree;
+
+/// \brief Enum for identifying different kinds of qualifiers.
+//
+// Notice that we are choosing to represent pointers and arrays as qualifiers.
+// The idea is that a qualifier is something that you can add to a type T to
+// obtain another type R, in such a way that if T is fully known also R is fully
+// known. In this sense Pointer and Array types are qualified types.
+namespace QualifierKind {
+enum Values { Invalid, Pointer, Array, Const };
+// TODO: fromString toString
+} // end namespace QualifierKind
+
+struct Qualifier;
+struct QualifiedType;
+
+struct EnumEntry;
+struct AggregateField;
+struct StructField;
+struct UnionField;
+struct ArgumentType;
+
+} // end namespace flatc
+
+// Make flatc::Type derived types usable with UpcastablePointer
+template<>
+struct concrete_types_traits<flatc::Type> {
+  using type = std::tuple<flatc::PrimitiveType,
+                          flatc::EnumType,
+                          flatc::TypedefType,
+                          flatc::StructType,
+                          flatc::UnionType,
+                          flatc::FunctionPtrType,
+                          flatc::Type>;
+};
+
+/// \brief Concept to identify all types that are derived from flatc::Type
+template<typename T>
+concept FlatCType = std::is_base_of_v<flatc::Type, T>;
+
+/// \brief Base class of flatc types used for LLVM-style RTTI
+struct flatc::Type {
+  static constexpr const char *Tag = "!flatc::InvalidType";
+
+  // TODO: we should switch to a sane UUID implementation
+  uint64_t UUID;
+
+  /// \brief The name of this type. This is the unique key to identifiy a type.
+  std::string Name;
+
+  /// \brief The kind of a flatc::Type
+  TypeKind::Values Kind;
+
+public:
+  TypeKind::Values getKind() const { return Kind; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Invalid;
+  }
+
+  Type(llvm::StringRef NameRef, TypeKind::Values TK) :
+    UUID(0ULL), Name(NameRef), Kind(TK) {
+    auto Err = llvm::getRandomBytes(&UUID, sizeof(UUID));
+    if (Err)
+      UUID = 0ULL;
+  }
+
+  Type(llvm::StringRef NameRef) : Type(NameRef, TypeKind::Values::Invalid) {}
+  Type() : Type("") {}
+  ~Type() = default;
+
+  Type(const Type &) = default;
+  Type &operator=(const Type &) = default;
+
+  Type(Type &&) = default;
+  Type &operator=(Type &&) = default;
```

Comment:
Why this again?

Reply:
Well, this is not strictly necessary right now. All the hierarchy of the types is defined in a way that the compiler will always be able to automatically default copy/move constructor/assignment and the destructor as well.

I explictly defaulted them mainly as a safeguard during development, when this classes were changing heavily and often. In this way whenever I managed to screw them up to a point where the compiler could not automatically default those, I would get a compilation error.

In the current state, I can drop all the explicitly defaulted destructors, and copy/move constructor/assignments and no functionality will change. We would just lose this safeguard in case of future development. In that case, at a certain point the compiler could stop being able to default those, but we would not get a compilation error. That's basically it.

Should I drop them?

Reply:
Let's drop everything can be dropped.
I favor simplicity and coherence with the status quo right now, if there's a benefit, we can introduce that in a distinct PR.

Reply:
Done